xcm/src/double_encoded.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use alloc::vec::Vec;
+use parity_scale_codec::{Encode, Decode};
+
+#[derive(Encode, Decode)]
+#[codec(encode_bound())]
+#[codec(decode_bound())]
+pub struct DoubleEncoded<T> {
+	encoded: Vec<u8>,
+	#[codec(skip)]
+	decoded: Option<T>,
+}
+
+impl<T> Clone for DoubleEncoded<T> {
+	fn clone(&self) -> Self { Self { encoded: self.encoded.clone(), decoded: None } }
```

Comment:
If you call `ensure_decoded` and then `clone` you dont want the decoded value?

```suggestion
	fn clone(&self) -> Self { Self { encoded: self.encoded.clone(), decoded: self.decoded.clone() } }
```